src/context.rs

```diff
@@ -111,32 +136,76 @@ impl<'a> QueryMatchContext<'a> {
     }
 }
 
-pub enum ParsedOrUnparsedQuery<'query_text> {
+pub enum ParsedOrUnparsedQuery<'a> {
     Parsed(Query),
-    Unparsed(&'query_text str),
+    ParsedRef(&'a Query),
+    Unparsed(&'a str),
 }
 
-impl<'query_text> ParsedOrUnparsedQuery<'query_text> {
-    pub fn into_parsed(self, language: Language) -> Query {
+impl<'a> ParsedOrUnparsedQuery<'a> {
+    pub fn parsed(&self, language: Language) -> MaybeOwned<'_, Query> {
+        match self {
+            Self::Parsed(query) => query.into(),
+            Self::ParsedRef(query) => (*query).into(),
+            Self::Unparsed(query_text) => Query::new(language, query_text).unwrap().into(),
+        }
+    }
+
+    pub fn into_parsed(self, language: Language) -> MaybeOwned<'a, Query> {
         match self {
-            Self::Parsed(query) => query,
-            Self::Unparsed(query_text) => Query::new(language, query_text).unwrap(),
+            Self::Parsed(query) => query.into(),
+            Self::ParsedRef(query) => query.into(),
+            Self::Unparsed(query_text) => Query::new(language, query_text).unwrap().into(),
         }
     }
 }
 
-impl<'query_text> From<Query> for ParsedOrUnparsedQuery<'query_text> {
+impl<'a> From<Query> for ParsedOrUnparsedQuery<'a> {
     fn from(value: Query) -> Self {
         Self::Parsed(value)
     }
 }
 
-impl<'query_text> From<&'query_text str> for ParsedOrUnparsedQuery<'query_text> {
-    fn from(value: &'query_text str) -> Self {
+impl<'a> From<&'a Query> for ParsedOrUnparsedQuery<'a> {
+    fn from(value: &'a Query) -> Self {
+        Self::ParsedRef(value)
+    }
+}
+
+impl<'a> From<&'a str> for ParsedOrUnparsedQuery<'a> {
+    fn from(value: &'a str) -> Self {
         Self::Unparsed(value)
     }
 }
 
+pub enum MaybeOwned<'a, T> {
```

Comment:
I can't remember if I ended up using this? If so I saw there's a `maybe_owned` crate version (same idea) and so would probably be preferable to use it

The basic idea is that `Cow` can only be used to express "maybe owned, maybe borrowed" for types that implement `ToOwned` (which `Clone` types do)

But so I was trying to express an "owned or borrowed" situation for something that wasn't `Clone` so just made a new enum (that implements `Deref`) to express that, which is basically what the `maybe_owned` crate says it's for too